### PR TITLE
fix(go): correctly handle void return types

### PR DESCRIPTION
**Pull request by MrArnoldPalmer**
Fixes go method code generation to not emit a return statement when the
return type is `void`. Adds logic to fix go compile errors on
getters/setters that refer to recursive type properties. Meaning, if a
struct contains a property that the type of the struct itself, the type
of the property in Go is a pointer to the struct. Getters and setters
now reflect this.

Update test step to include compilation of all generated jsii-calc
modules since all compile errors are fixed.

fix:  #2009

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/jsii-pacmak/lib/targets/go/types/class.ts [hunk 1]
```diff
@@ -140,9 +140,10 @@ export class ClassMethod implements GoTypeMember {
     const name = this.name;
 
     const instanceArg = this.parent.name.substring(0, 1).toLowerCase();
+    const returnTypeString = this.reference?.void ? '' : ` ${this.returnType}`;
```
Comment (SoManyHs) [inline packages/jsii-pacmak/lib/targets/go/types/class.ts:143]: any reason not to incorporate this logic in the `returnType` getter? That was the point of https://github.com/aws/jsii/pull/2001

File: packages/jsii-pacmak/lib/targets/go/types/go-type.ts [hunk 2]
```diff
@@ -189,7 +189,12 @@ export class GoProperty implements GoTypeMember {
         this.getter
       }()${` ${this.returnType}`}`,
     );
-    code.line(`return ${instanceArg}.${this.name}`);
+
+    if (this.parent.name === this.returnType) {
+      code.line(`return *${instanceArg}.${this.name}`);
+    } else {
+      code.line(`return ${instanceArg}.${this.name}`);
```
Comment (SoManyHs) [inline packages/jsii-pacmak/lib/targets/go/types/go-type.ts:196]: Wait, I'm confused by this -- why is this necessary on emitting the Getter implementation?

File: packages/jsii-pacmak/lib/targets/go/types/go-type.ts [hunk 3]
```diff
@@ -202,7 +207,12 @@ export class GoProperty implements GoTypeMember {
     code.openBlock(
       `func (${instanceArg} *${receiver}) Set${this.name}(val ${this.returnType})`,
     );
-    code.line(`${instanceArg}.${this.name} = val`);
+
+    if (this.parent.name === this.returnType) {
+      code.line(`${instanceArg}.${this.name} = &val`);
+    } else {
+      code.line(`${instanceArg}.${this.name} = val`);
+    }
```
Comment (SoManyHs) [inline packages/jsii-pacmak/lib/targets/go/types/go-type.ts:215]: same as above?

File: packages/jsii-pacmak/test/generated-code/__snapshots__/target-go.test.ts.snap [hunk 4]
```diff
@@ -7443,7 +7320,7 @@ func (s *Statics) GetZooBar() map[string]string {
 }
 
 func (s *Statics) GetInstance() Statics {
-    return s.Instance
+    return *s.Instance
```
Comment (MrArnoldPalmer) [inline packages/jsii-pacmak/test/generated-code/__snapshots__/target-go.test.ts.snap:7323]: @SoManyHs here is an example of the change to getters. This may only be necessary as long as the returns of getters are direct property accesses and not calls to the JSII runtime. But for now this is necessary because the `Instance` property of statics is a `*static` so we have to deref it to make it the proper return type.

File: packages/jsii-pacmak/lib/targets/go/types/go-type.ts (hunk 2 above)
Comment (MrArnoldPalmer) [reply packages/jsii-pacmak/lib/targets/go/types/go-type.ts:196]: See https://github.com/aws/jsii/pull/2035/files#r492994719

File: packages/jsii-pacmak/lib/targets/go/types/class.ts (hunk 1 above)
Comment (MrArnoldPalmer) [reply packages/jsii-pacmak/lib/targets/go/types/class.ts:143]: The main reason was consistency. To only add spacing to the class returnType which affects how others call it, [for example the runtime call](https://github.com/aws/jsii/blob/master/packages/jsii-pacmak/lib/targets/go/runtime.ts#L34). I think `returnType` should stay generic enough to be useful outside of the class and not concerned with formatting.

**Review by SoManyHs — APPROVED** [review]
